### PR TITLE
[enhance](PrefetchReader) abort load task when data size returned by S3 is smaller than requested

### PR DESCRIPTION
**Pull request by ByteYue**
# Proposed changes

Issue Number: close #xxx
We encountered one confusing situation where buffered reader were trapped in one endless loop when calling readat. Then we found out that it was all due to the return data size is less than requested.
As the following picture shows, the actual data size is about 2M, and when we called readat it only retrieved about 1MB.
image

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run beut

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.21 seconds
 stream load tsv:          448 seconds loaded 74807831229 Bytes, about 159 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          58 seconds loaded 1101869774 Bytes, about 18 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          79.0 seconds inserted 10000000 Rows, about 126K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230523041332_clickbench_pr_148700.html

**Comment by AshinGau** [comment]
LGTM

File: be/src/io/fs/buffered_reader.cpp [hunk 1]
```diff
@@ -410,6 +411,13 @@ void PrefetchBuffer::prefetch_buffer() {
         SCOPED_RAW_TIMER(&_statis.read_time);
         s = _reader->read_at(_offset, Slice {_buf.get(), buf_size}, &_len, _io_ctx);
     }
+    if (UNLIKELY(buf_size != _len)) {
+        // This indicates that the data size returned by S3 object storage is smaller than what we requested,
+        // which seems to be a violation of the S3 protocol since our request range was valid.
+        // We currently consider this situation a bug and will treat this task as a failure.
+        s = Status::InternalError("Data size returned by S3 is smaller than requested");
```
Comment (morningman) [inline be/src/io/fs/buffered_reader.cpp:418]: Print details? like buf_size and _len

**Comment by ByteYue** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.77 seconds
 stream load tsv:          450 seconds loaded 74807831229 Bytes, about 158 MB/s
 stream load json:         22 seconds loaded 2358488459 Bytes, about 102 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          76.8 seconds inserted 10000000 Rows, about 130K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230526045620_clickbench_pr_151042.html

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**